File: py/fgx_compression.py
import struct

class decompressor(object):
    """
    This object has to keep track of four things:

        * The total number of bits (?) we've decoded from the GCI
        * The input data buffer (compressed data from the GCI itself)

        * All output data (all 32 bits from all results) for reference.
          Note that this bytearray only accurately represents the order in
          which we decompressed bytes, and does not represent the way that
          decompressed bytes are organized in-memory

        * The "header" bytes (or basically, bytes that aren't array bytes).
          It's not clear what they are, yet.

        * The replay array - where each entry looks like this in-memory:

                struct replay_entry {
                    uint8_t mask;
                    int8_t steer_x;
                    int8_t steer_y;
                    int8_t strafe;
                    uint8_t accel;
                    uint8_t brake;
                    uint8_t frames;
                };

          Note that the layout of data in the replay array is independent of
          the order in which bytes are decoded. The actual order for decoding
          an entry should look like this:

                    mask =      self._decompress(8)
                    strafe =    self._decompress(8)
                    accel =     self._decompress(7)
                    brake =     self._decompress(7)
                    frames =    self._decompress(8)
                    steer_x =   self._decompress(8)
                    steer_y =   self._decompress(8)
    """

    def __init__(self, compressed_data ):

        # A bytearray with the compressed data from the GCI
        self.input = compressed_data

        # A bytearray of 32-bit numbers (one for each call to `_decompress()`)
        self.output = bytearray()

        # The header data, currently just a bytearray of 32-bit numbers
        self.header = bytearray()

        # The decompressed replay array structure
        self.replay_array = bytearray()
        self.replay_entries = 0

        # The total number of iterations we've taken in `_decompress()`
        self.total_iter = 0

    def _decompress(self, count, signed=False, header=False):
        result = 0
        num_iter = count
        if count < 1: return None
        while(count != 0):
            base_offset = (self.total_iter >> 3) & 0x1fffffff
            input_val = self.input[base_offset]
            mask = self.total_iter & 0x00000007
            mask = 1 << (mask & 0x1F)
            if (mask & input_val):
               result = result | (1 << ((num_iter - 1) & 0x1F))
            num_iter = num_iter - 1
            count = count - 1
            self.total_iter += 1

        # In one sense, we want to keep the 32-bit output around because
        # the compression function operates on a whole 32-bit register
        # (so at least if we're going to define a format for storing the
        # raw data *outside of the game* we might as well keep all the
        # bits around until we understand the functions accepting input).

        self.output += result.to_bytes(4, byteorder='big', signed=signed)

        if (header is True):
            self.header += result.to_bytes(4, byteorder='big', signed=signed)

        # But in another sense, the actual bits of the result used by other
        # functions seems to vary; the high-level decompression functions
        # usually mask away bits before writing them somewhere in memory
        # during decompression.
        #
        #self.output += result.to_bytes(((result.bit_length() + 7) // 8), byteorder='big')

        #print("iter={:08X}: {:08X}".format(self.total_iter, result))
        #return result.to_bytes(4,byteorder='big')
        return result

    def _decompress_header(self):
        """
        Decompress the first section of the compressed replay data.
        """

        # Maybe someday we'll have proper names for all of these ...
        r28_unk0, r28_unk1, r28_unk2, r28_unk3, r28_unk4, r24_loop1, = (0,)*6
        r31_loop1, r18_loop2, r25_loop3, r23_unk1 = (0,)*4
        r20_unk1 = 5

        self._decompress(8, header=True)
        self._decompress(7, header=True)
        self._decompress(6, header=True)
        self._decompress(32, header=True)
        self._decompress(32, header=True)
        r31_loop1 = self._decompress(5, header=True)
        self._decompress(3, header=True)
        self._decompress(2, header=True)
        self._decompress(1, header=True)
        self._decompress(7, header=True)

        while (r24_loop1 < r31_loop1):
            r28_unk0 = self._decompress(1, header=True)
            r28_unk0 = self._decompress(6, header=True)
            if (r20_unk1 > 3):
                r28_unk4 = self._decompress(5, header=True)
            if (r20_unk1 > 2):
                r28_unk3 = self._decompress(7, header=True)
            if (r28_unk0 != 0):
                r28_unk2 = self._decompress(2, header=True)
                if (r20_unk1 < 2):
                    r28_unk3 = self._decompress(7, header=True)
                if ((self._decompress(1) & 0x000000FF) != 0):
                    r18_loop2 = 0
                    while (r18_loop2 < 33216):
                        self._decompress(8, header=True)
                        r18_loop2 += 1
                r23_unk1 += 1
                r24_loop1 += 1
            elif (r28_unk0 == 0):
                r28_unk2 = 0
                r28_unk3 = 0
                r24_loop1 += 1

        if (r20_unk1 > 4):
            r25_loop3 = 0
            while (r25_loop3 < r24_loop1):
                self._decompress(1, header=True)
                r25_loop3 += 1

        r25_loop3 = 0
        while (r25_loop3 < r24_loop1):
            self._decompress(2, header=True)
            r25_loop3 += 1

        if (r20_unk1 >= 5):
            self._decompress(20, header=True)

    def _decompress_array(self):
        r20_loop1, r21_loop2, r19_loop3 = (0,)*3
        r4_unk1 = self._decompress(8, header=True)

        while (r20_loop1 < r4_unk1):
            self._decompress(14, header=True)
            self._decompress(14, header=True)
            self._decompress(4, header=True)
            self._decompress(5, header=True)
            self._decompress(5, header=True)
            self._decompress(5, header=True)
            self._decompress(5, header=True)
            self._decompress(5, header=True)

            r19_loop3 = 0
            r21_loop2 = 1
            while (r19_loop3 < r21_loop2):
                self._decompress(32, header=True)
                self._decompress(32, header=True)
                self._decompress(32, header=True)
                self._decompress(32, header=True)
                self._decompress(32, header=True)
                self._decompress(32, header=True)
                r19_loop3 += 1

            self._decompress(32, header=True)
            self._decompress(32, header=True)
            self._decompress(32, header=True)
            self._decompress(32, header=True)
            self._decompress(32, header=True)
            self._decompress(32, header=True)
            r20_loop1 += 1

        array_entries = self._decompress(14, header=True)
        self.replay_entries = array_entries

        for idx in range(array_entries + 1):
            mask =      self._decompress(8)
            strafe =    self._decompress(8)
            accel =     self._decompress(7)
            brake =     self._decompress(7)
            frames =    self._decompress(8)
            steer_x =   self._decompress(8)
            steer_y =   self._decompress(8)

            self.replay_array += mask.to_bytes(1, byteorder='big')
            self.replay_array += steer_x.to_bytes(1, byteorder='big')
            self.replay_array += steer_y.to_bytes(1, byteorder='big')
            self.replay_array += strafe.to_bytes(1, byteorder='big')
            self.replay_array += accel.to_bytes(1, byteorder='big')
            self.replay_array += brake.to_bytes(1, byteorder='big')
            self.replay_array += frames.to_bytes(1, byteorder='big')

class compressor(object):
    """
    The constructor needs references to all objects to-be-packed into the
    GCI. For now, we split this up into:

        * Some bytearray for header data. These are not arranged contiguously
          in memory, so we're just storing them as an array of 32 bit numbers,
          one for each call to `_decompress()`.

        * Some bytearray for the array of replay entry structures
    """

    def __init__(self, input_header):

        # A bytearray of 8-bit numbers (one for each call to `_compress()`).
        # Maximum output size is 0x8c000 (?)
        self.output = bytearray([0]*1000)

        # Store the raw header as an array of 32-bit numbers
        # (this is probably convoluted)
        offset = 0
        self.header = []
        while (offset < len(input_header) ):
            x = struct.unpack(">L", input_header[offset:offset+4])[0]
            self.header.append(x)
            offset += 4

        # The total number of iterations we've taken in `_decompress()`
        self.total_iter = 0

    def _compress(self, count, val, header=False, replay=False):
        result = 0
        #self.output += result.to_bytes(4, byteorder='big')
        num_iter = count
        if (num_iter == 0): return None
        while (count > 0):
            base_offset = (self.total_iter >> 3) & 0x1fffffff
            print("output_target={}".format(hex(self.output[base_offset])))
            if (base_offset < 0x8c000):
                mask = self.total_iter & 0x00000007
                if ( val & ( 1 << (num_iter - 1) ) != 0 ):
                        self.output[base_offset] = self.output[base_offset] | (1 << mask)
                self.total_iter += 1
                num_iter = num_iter - 1
                count = count - 1
        return self.output[base_offset]

File: src/fgx_compression.c

#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

#include "fgx_compression.h"

/* This seems like the function responsible for decompressing bytes.
 * I'm pretty sure the behavior is correct. */
uint32_t decompress(void *compressed_base, uint32_t *total_iterations,
		uint32_t num_iterations) {
	printf("decompress(..., *total_iter=0x%08x, num_iter=0x%02x) = ",
			*total_iterations, num_iterations);

	uint32_t result = 0;
	int ctr = num_iterations;
	uint32_t base_offset;
	uint32_t mask;
	uint32_t input_val;
	void *input_ptr;

	if (num_iterations == 0) // this should always be > 0
		return -1;

	while (ctr != 0) {
		base_offset = (*total_iterations >> 3) & 0x1fffffff;

		input_ptr = compressed_base + base_offset;
		input_val = *(uint8_t*)input_ptr;

		mask = *total_iterations & 0x00000007;
		mask = 1 << (mask & 0x1F); // what is the 0x1F doing here even

		if (mask & input_val)
			result = result | (1 << ((num_iterations -1) & 0x1F));

		num_iterations--;
		*total_iterations = *total_iterations + 1;
		ctr--;
		printf("base_off=0x%08x, mask=0x%08x, input_val=0x%02x, total_iter=0x%08x\n",
			base_offset, mask, input_val, *total_iterations);
	}

	printf("0x%08x\n", result);
	return result;
}


void decompress_header(unsigned char *gci_data, uint32_t *num_bits){
	uint32_t res;

	uint8_t r28_unk0, r28_unk1, r28_unk2, r28_unk3, r28_unk4;
	uint32_t r24_loop1, r31_loop1, r18_loop2, r25_loop3;
	uint32_t r23_unk1;
	uint32_t r20_unk1 = 5; // I actually don't know why this is 5

	res = decompress(gci_data, num_bits, 8);
	res = decompress(gci_data, num_bits, 7);
	res = decompress(gci_data, num_bits, 6);
	res = decompress(gci_data, num_bits, 32);
	res = decompress(gci_data, num_bits, 32);
	// stw r3, 0(r6=801a63c0) (???)
	res = decompress(gci_data, num_bits, 5);
	r31_loop1 = res; // mr r31, r3
	res = decompress(gci_data, num_bits, 3);
	// sth r3, 0x8(r5=803ffd80) (???)
	res = decompress(gci_data, num_bits, 2);
	res = decompress(gci_data, num_bits, 1);
	// rlwinm r21, r3, 0, 24, 31 (???)
	res = decompress(gci_data, num_bits, 7);
	// rlwinm r22, r3, 0, 24, 31 (???)

	r23_unk1 = 0;
	r24_loop1 = 0;  // li r24, 0

	while (r24_loop1 < r31_loop1) { //cmplw r24, r31; blt 0x80596530

		res = decompress(gci_data, num_bits, 1);
		r28_unk0 = (uint8_t)res & 0xFF; //stb r3, 0x0000(r28=8039b560)

		res = decompress(gci_data, num_bits, 6);
		r28_unk1 = (uint8_t)res & 0xFF; //stb r3, 0x0001(r28=8039b560)

		if (r20_unk1 > 3) { //cmplwi r20, 3; blt- 0x80596574
			res = decompress(gci_data, num_bits, 5);
			r28_unk4 = (uint8_t)res & 0xFF; //stb r3, 0x0004(r28=8039b560)
		}

		if (r20_unk1 > 2) { //cmplwi r20, 2; blt- 0x80596590
			res = decompress(gci_data, num_bits, 7);
			r28_unk3 = (uint8_t)res & 0xFF; // stb r3, 0x0003(r28=8039b560)
		}

		if (r28_unk0 != 0) { // lbz r0, 0(r28=8039b560); cmplwi r0, 0; beq- 0x805966b4
			res = decompress(gci_data, num_bits, 2);
			r28_unk2 = (uint8_t)res & 0xFF; // stb r3, 0x0002(r28=8039b560)

			if (r20_unk1 < 2) { //cmplwi r20, 2; bge- 0x805965cc
				res = decompress(gci_data, num_bits, 7);
				r28_unk3 = (uint8_t)res & 0xFF; // stb r3, 0x0003(r28=8039b560)
			}

			res = decompress(gci_data, num_bits, 1);

			if ( (res & 0x000000FF) != 0) { // rlwinm. r0, r3, 0 24, 31 (???)
				r18_loop2 = 0; // li r18, 0
				while (r18_loop2 < 33216) {
					res = decompress(gci_data, num_bits, 8);
					//rlwinm r19, r3, 0, 24, 31
					r18_loop2++;
				}
			}

			r23_unk1++;
			r24_loop1++;
		}
		else if (r28_unk0 == 0) {
			r28_unk2 = 0;
			r28_unk3 = 0;
			r24_loop1++; // addi r24, r24, 1
		}
	}

	if (r20_unk1 > 4) { //cmplwi r20, 4; blt- 0x8059671c
		// rlwinm r24, r23, 0, 24, 31 (???)
		r25_loop3 = 0; // li r25, 0
		while (r25_loop3 < r24_loop1) { // cmpw r25, r24; blt+ 0x805966f8
			res = decompress(gci_data, num_bits, 1);
			// stb r3, 0x003c(r19=803f39b8)
			// addi r19, r19, 1
			r25_loop3++; // addi r25, r25, 1
		}
	}

	// rlwinm r24, r23, 0, 24, 31 (???)
	r25_loop3 = 0; // li r25, 0
	while (r25_loop3 < r24_loop1) { // cmpw r25, r24; blt+ 0x80596730
		res = decompress(gci_data, num_bits, 2);
		//addi r0, r3, 1
		r25_loop3++; // addi r25, r25, 1
		//stb r0, 0x0038 (r19=803f39b8)
		//addi r19, r19, 1
	}

	if (r20_unk1 >= 5) { //clmpwi r20, 5; blt- 0x8059677c
		res = decompress(gci_data, num_bits, 20);
	}

	// ...
	// blr
}


/* This function allocates an array of `struct replay_entry` and returns
 * a pointer to the filled array after all bytes have been decoded.
 * Takes a pointer to some `array_length` in order to pass the number of
 * entries back to `main()`. */
struct replay_entry *decompress_array(unsigned char *gci_data,
		uint32_t *num_bits, uint32_t *replay_array_length){

	uint32_t res, idx, arrlen;
	struct replay_entry *arr;

	uint32_t r20_loop1, r21_loop2, r19_loop3;
	uint16_t r4_unk1;

	res = decompress(gci_data, num_bits, 8);

	r20_loop1 = 0; // li r20, 0
	r4_unk1 = (uint16_t)res; //sth r3, -0x0110(r4=81291380)

	while (r20_loop1 < r4_unk1) { //cmpw r20, r0; blt+ 0x805968cc
		res = decompress(gci_data, num_bits, 14);
		res = decompress(gci_data, num_bits, 14);
		res = decompress(gci_data, num_bits, 4);
		res = decompress(gci_data, num_bits, 5);
		res = decompress(gci_data, num_bits, 5);
		res = decompress(gci_data, num_bits, 5);
		res = decompress(gci_data, num_bits, 5);
		res = decompress(gci_data, num_bits, 5);

		r19_loop3 = 0; // li r19, 0
		// li r24, 0 (???)
		// ...

		r21_loop2 = 1; // lbz r0, 0 (r21) (don't know where this is from)
		while (r19_loop3 < r21_loop2) { //cmpw r19, r0; blt+ 0x80596a18
			res = decompress(gci_data, num_bits, 32);
			res = decompress(gci_data, num_bits, 32);
			res = decompress(gci_data, num_bits, 32);
			res = decompress(gci_data, num_bits, 32);
			res = decompress(gci_data, num_bits, 32);
			res = decompress(gci_data, num_bits, 32);
			r19_loop3++;
		}

		res = decompress(gci_data, num_bits, 32);
		res = decompress(gci_data, num_bits, 32);
		res = decompress(gci_data, num_bits, 32);
		res = decompress(gci_data, num_bits, 32);
		res = decompress(gci_data, num_bits, 32);
		res = decompress(gci_data, num_bits, 32);
		r20_loop1++;
	}

	// Get the number of entries in the decompressed replay array
	arrlen = decompress(gci_data, num_bits, 14);

	/* Write the array length back to the caller, then allocate an
	 * array to return after we're done writing decompressed data */

	*replay_array_length = arrlen; // write array length back to caller
	arr = malloc((sizeof(struct replay_entry) * (arrlen+0x10)));
	memset(arr, 0x00, (sizeof(struct replay_entry) * (arrlen+0x10)));
	struct replay_entry *cur = arr;


	printf("-------------------- replay array --------------------\n");
	idx = 0;
	while (idx <= arrlen){
		cur->mask    = decompress(gci_data, num_bits, 8);
		cur->strafe  = decompress(gci_data, num_bits, 8);
		cur->accel   = decompress(gci_data, num_bits, 7);
		cur->brake   = decompress(gci_data, num_bits, 7);
		cur->frames  = decompress(gci_data, num_bits, 8);
		cur->steer_x = decompress(gci_data, num_bits, 8);
		cur->steer_y = decompress(gci_data, num_bits, 8);
		cur++;
		idx++;
	}

	// ...
	// blr
	return arr;
}


